x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/persistence/TrainedModelProvider.java

```diff
@@ -1053,63 +1058,84 @@ public void expandIds(
+                client.threadPool().getThreadContext(),
+                ML_ORIGIN,
+                multiSearchRequest,
+                ActionListener.<MultiSearchResponse>wrap(responses -> {
+                    List<InferenceStats> allStats = new ArrayList<>(modelIds.length);
+                    int modelIndex = 0;
+                    assert responses.getResponses().length == modelIds.length
+                        : "mismatch between search response size and models requested";
+                    for (MultiSearchResponse.Item response : responses.getResponses()) {
+                        if (response.isFailure()) {
+                            if (ExceptionsHelper.unwrapCause(response.getFailure()) instanceof ResourceNotFoundException) {
+                                modelIndex++;
+                                continue;
+                            }
+                            logger.error(
+                                () -> "[" + Strings.arrayToCommaDelimitedString(modelIds) + "] search failed for models",
+                                response.getFailure()
+                            );
+                            listener.onFailure(
+                                ExceptionsHelper.serverError(
+                                    "Searching for stats for models [{}] failed",
+                                    response.getFailure(),
+                                    Strings.arrayToCommaDelimitedString(modelIds)
+                                )
+                            );
+                            return;
                         }
-                        logger.error(
-                            () -> "[" + Strings.arrayToCommaDelimitedString(modelIds) + "] search failed for models",
-                            response.getFailure()
-                        );
-                        listener.onFailure(
-                            ExceptionsHelper.serverError(
-                                "Searching for stats for models [{}] failed",
-                                response.getFailure(),
-                                Strings.arrayToCommaDelimitedString(modelIds)
-                            )
-                        );
-                        return;
-                    }
-                    try {
-                        InferenceStats inferenceStats = handleMultiNodeStatsResponse(response.getResponse(), modelIds[modelIndex++]);
-                        if (inferenceStats != null) {
-                            allStats.add(inferenceStats);
+                        try {
+                            InferenceStats inferenceStats = handleMultiNodeStatsResponse(response.getResponse(), modelIds[modelIndex++]);
+                            if (inferenceStats != null) {
+                                allStats.add(inferenceStats);
+                            }
+                        } catch (Exception e) {
+                            listener.onFailure(e);
```

Comment:
I think we want to call `delegate` from this scope, and it will in turn call the listener added in `addListener`?

Reply:
yea, we can definitely do that. I didn't bother to change it because I don't think it has any effect, but I'm happy to do it if you prefer it.